### PR TITLE
LLVM: Implement associated and null

### PR DESCRIPTION
**Pull request by Smit-create**
Extending #1152 to LLVM

### CODE REVIEW

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 1]
```diff
@@ -4312,6 +4322,36 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
...
+        ASR::dimension_t* m_dims = nullptr;
+        int n_dims = -1, a_kind = -1;
+        ptr = CreateLoad(ptr);
+        llvm::Type* value_type = get_type_from_ttype_t(p->m_type, m_storage, is_array_type,
+                                    is_malloc_array_type, is_list, m_dims, n_dims, a_kind);
+        if (x.m_tgt) {
+            ASR::Variable_t *t = EXPR2VAR(x.m_tgt);
+            uint32_t t_h = get_hash((ASR::asr_t*)t);
+            llvm::Value *nptr = llvm_symtab[t_h];
+            llvm::Value *diff = builder->CreatePtrDiff(value_type, ptr, nptr);
```
Comment (Smit-create) [inline src/libasr/codegen/asr_to_llvm.cpp:4346]: The CI fails here due to API changes in LLVM 14 and earlier versions.

Comment (Smit-create) [reply src/libasr/codegen/asr_to_llvm.cpp:4346]: See--LLVM 10 fails but LLVM 14+ passes with this diff.

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 2]
```diff
@@ -4312,6 +4322,36 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
...
+            tmp = builder->CreateICmpEQ(diff, const_zero);
+        } else {
+            llvm::Value *nptr = llvm::ConstantPointerNull::get(static_cast<llvm::PointerType*>(value_type));
+            llvm::Value *diff = builder->CreatePtrDiff(value_type, ptr, nptr);
```
Comment (certik) [inline src/libasr/codegen/asr_to_llvm.cpp:4350]: Create a function `CreatePtrDiff(builder, value_type, ptr, nptr)` in `llvm_utils.h` and implement it using:
```c++
#if LLVM_VERSION_MAJOR >= 14
return builder->CreatePtrDiff(value_type, ptr, nptr)
#else                              
builder->CreatePtrDiff(ptr, nptr)
#endif                                       
```

**Comment by certik** [comment]
Otherwise it looks good.

**Review by certik — APPROVED** [review]
Looks great!

**Comment by certik** [comment]
You'll have to fix this one:
```
/home/runner/work/lfortran/lfortran/src/libasr/codegen/llvm_utils.cpp:30:76: error: unused parameter ‘type’ [-Werror=unused-parameter]
   30 |         llvm::Value* CreatePtrDiff(llvm::IRBuilder<> &builder, llvm::Type *type,
      |                                                                ~~~~~~~~~~~~^~~~
```
By using ifdef there as well: `llvm::Type* /*type*/`

File: src/libasr/codegen/llvm_utils.h [hunk 3]
```diff
@@ -81,6 +81,8 @@ namespace LCompilers {
 
         llvm::Value* CreateLoad(llvm::IRBuilder<> &builder, llvm::Value *x);
         llvm::Value* CreateStore(llvm::IRBuilder<> &builder, llvm::Value *x, llvm::Value *y);
+        llvm::Value* CreatePtrDiff(llvm::IRBuilder<> &builder, llvm::Type *type,
```
Comment (czgdp1807) [inline src/libasr/codegen/llvm_utils.h:84]: Why do we need the difference between two pointer values? How is it helping in the implementations added in this PR?

Comment (Smit-create) [reply src/libasr/codegen/llvm_utils.h:84]: To check if the pointer is pointing to the given target by calculating the difference in the address values. If it's 0 then they are associated or else not.

Comment (czgdp1807) [reply src/libasr/codegen/llvm_utils.h:84]: You can just to simple comparison by converting pointer to int? `CreatePtrToInt` and `CreateICmpEq` I guess. Why is the difference necessary here when a simple comparison can do the same job? Is the value of the difference important if its non-zero? Do we use it somewhere else?

Comment (Smit-create) [reply src/libasr/codegen/llvm_utils.h]: Fixed that using `CreatePtrToInt`

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 4]
```diff
@@ -4315,6 +4325,35 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
...
+            uint32_t t_h = get_hash((ASR::asr_t*)t);
+            nptr = llvm_symtab[t_h];
+            nptr = builder->CreatePtrToInt(nptr, getIntType(8, false));
+            tmp = builder->CreateICmpEQ(ptr, nptr);
+        } else {
+            ASR::storage_typeType m_storage = ASR::storage_typeType::Default;
+            bool is_array_type = false, is_malloc_array_type = false, is_list = false;
+            ASR::dimension_t* m_dims = nullptr;
+            int n_dims = -1, a_kind = -1;
+            llvm::Type* value_type = get_type_from_ttype_t(p->m_type, m_storage, is_array_type,
```
Comment (czgdp1807) [inline src/libasr/codegen/asr_to_llvm.cpp:4349]: You can use, `get_type_from_ttype_t_util`.

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 5]
```diff
@@ -2392,6 +2392,16 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
         }
     }
 
+    void visit_PointerNullConstant(const ASR::PointerNullConstant_t& x){
+        ASR::storage_typeType m_storage = ASR::storage_typeType::Default;
+        bool is_array_type = false, is_malloc_array_type = false, is_list = false;
+        ASR::dimension_t* m_dims = nullptr;
+        int n_dims = -1, a_kind = -1;
+        llvm::Type* value_type = get_type_from_ttype_t(x.m_type, m_storage, is_array_type,
+                                    is_malloc_array_type, is_list, m_dims, n_dims, a_kind);
```
Comment (czgdp1807) [inline src/libasr/codegen/asr_to_llvm.cpp:2401]:
```suggestion
        llvm::Type* value_type = get_type_from_ttype_t_util(x.m_type);
```

**Review by czgdp1807 — APPROVED** [review]
Other than the suggestion. LGTM. Once done, mark it ready and merge.

**Pull request merged**